.changeset/strange-kids-change.md

```diff
@@ -0,0 +1,5 @@
+---
+'@emotion/styled': patch
+---
+
+Fix hydration mismatches if `React.useId` is used
```

Comment:
I wonder - would you be open to creating a PR targeting this PR that would add tests for this (this would require bumping React version)? I would love to have a way to verify changes like this but I'm also hesitant to land React 18 on our main branch before it hits its stable mark.

Reply:
A robust hydration test suite is pretty involved since you really need two separated processes: one for SSR and one for hydrating. 

> I would love to have a way to verify changes like this but I'm also hesitant to land React 18 on our main branch before it hits its stable mark.

The linked codesandbox in the issue description should be sufficient for manual verification. It has a hydration mismatch with the latest version on npm but none with the codesandbox build from this PR:

Broken with `@emotion/styled@11.3.0`: https://codesandbox.io/s/sad-breeze-qznnl?file=/src/Sidebar.js
Fixed with this PR: https://codesandbox.io/s/serene-platform-r4eyr?file=/src/Sidebar.js

Reply:
> A robust hydration test suite is pretty involved since you really need two separated processes: one for SSR and one for hydrating.

Ye, I'm not for something like that now. It's possible to just mock whatever `isBrowser` checks before server and client renders to get a good representation of the real situation. For example [this PR](https://github.com/emotion-js/emotion/pull/2334) has introduced such tests [here](https://github.com/emotion-js/emotion/pull/2334/files#diff-8a41a3bb909094190325495242aab95bc553c6769d9060da0f18c4d3fafcc86d) (IIRC)

Reply:
Ah, thanks for the pointers. Will look into it.

> I would love to have a way to verify changes like this but I'm also hesitant to land React 18 on our main branch before it hits its stable mark.

I can install it behind an alias and let jest use it for specific tests.

Reply:
Oh, that would be cool - I have never used aliases like this. Hopefully, this won't be too problematic ;p Having that in place would be awesome ❤️

Reply:
Test illustrating the behavior without this change: [`862729d` (#2542)](https://github.com/emotion-js/emotion/pull/2542/commits/862729d103ebd78791198191201c6f42a9cd356f)